query-algebrizer/src/clauses/or.rs

```diff
@@ -678,6 +693,46 @@ impl ConjoiningClauses {
     }
 }
 
+/// Helper to fold together a set of type maps.
+fn union_types(into: &mut BTreeMap<Variable, ValueTypeSet>,
+               additional_types: &BTreeMap<Variable, ValueTypeSet>) {
+    // We want the exclusive disjunction -- any variable not mentioned in both sets -- to default
+    // to ValueTypeSet::Any.
+    // This is necessary because we lazily populate known_types, so sometimes the type set will
+    // be missing a `ValueTypeSet::Any` for a variable, and we want to broaden rather than
+    // accidentally taking the other side's word for it!
+    // The alternative would be to exhaustively pre-fill `known_types` with all mentioned variables
```

Comment:
This is complicated because our "type map" isn't Python's `defaultdict`.  Why not make it so?  (Assuming we find a Rust implementation.)  Of course, we could also produce our own newtype for this.